symbolic-debuginfo/src/sourcebundle.rs

```diff
@@ -173,10 +181,107 @@ pub struct SourceFileInfo {
     #[serde(default, skip_serializing_if = "String::is_empty")]
     url: String,
 
-    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
+    #[serde(
+        default,
+        skip_serializing_if = "BTreeMap::is_empty",
+        deserialize_with = "deserialize_headers"
+    )]
     headers: BTreeMap<String, String>,
 }
 
+/// A descriptor that provides information about a source file.
+///
+/// This descriptor is returned from [`source_by_path`](DebugSession::source_by_path)
+/// and friends.
+///
+/// This descriptor holds information that can be used to retrieve information
+/// about the source file.  A descriptor has to have at least one of the following
+/// to be valid:
+///
+/// - [`contents`](Self::contents)
+/// - [`remote_url`](Self::remote_url)
+/// - [`debug_id`](Self::debug_id)
+///
+/// Debug sessions are not permitted to return invalid source file descriptors.
+pub struct SourceFileDescriptor<'a> {
+    contents: Option<Cow<'a, str>>,
+    remote_url: Option<Cow<'a, str>>,
+    file_info: Option<&'a SourceFileInfo>,
+}
+
+impl<'a> SourceFileDescriptor<'a> {
+    /// Creates an embedded source file descriptor.
+    pub(crate) fn new_embedded(
+        content: Cow<'a, str>,
+        file_info: Option<&'a SourceFileInfo>,
+    ) -> SourceFileDescriptor<'a> {
+        SourceFileDescriptor {
+            contents: Some(content),
+            remote_url: None,
+            file_info,
+        }
+    }
+
+    /// Creates an remote source file descriptor.
+    pub(crate) fn new_remote(remote_url: Cow<'a, str>) -> SourceFileDescriptor<'a> {
+        SourceFileDescriptor {
+            contents: None,
+            remote_url: Some(remote_url),
+            file_info: None,
+        }
+    }
+
+    /// The type of the file the descriptor points to.
+    pub fn ty(&self) -> SourceFileType {
+        self.file_info
+            .and_then(|x| x.ty())
+            .unwrap_or(SourceFileType::Source)
+    }
+
+    /// The contents of the source file as string, if it's available.
+    pub fn contents(&self) -> Option<&str> {
+        self.contents.as_deref()
+    }
+
+    /// The remote URL that contains the source.
+    ///
+    /// There are cases where the descriptor itself is a reference in which case
+    /// an external service needs to be consulted to retrieve the information.
+    pub fn remote_url(&self) -> Option<&str> {
+        self.remote_url.as_deref()
+    }
+
+    /// The debug ID of the file if available.
+    pub fn debug_id(&self) -> Option<DebugId> {
+        self.file_info.and_then(|x| x.debug_id())
+    }
+
+    /// The source mapping URL reference of the file.
+    pub fn source_mapping_url(&self) -> Option<&str> {
```

Comment:
In theory this function has enough information to also parse out the `sourceMappingURL` comment in the contents. Not sure if it's a good idea to put it there, but it could be made possible. Whatever we end up doing, this probably should document the limitations of this reference.